### DISCUSSION TITLE
[React-ARIA] - Maximum update depth exceeded while building popover component with render props

### DISCUSSION THREAD

**Discussion by bstro** [adobe/react-spectrum#2450] [Q&A] [answered]
I'm stumped! Looking for a little help here, not sure if I'm doing something wrong, or if this is a bug.

I'm trying to build a flexible Popover component by using the [example for useOverlayTrigger](https://react-spectrum.adobe.com/react-aria/useOverlayTrigger.html#example) as a jumping-off point. I am trying to make this a flexible component for other developers on my team and I'm attempting to use the render-props pattern so that other developers don't have to worry about hooking up all the React-ARIA stuff in order to use the Popover. Here's a simplified reproduction in [CodeSandbox](https://codesandbox.io/s/nervous-murdock-fi2x4?file=/src/App.tsx).

Upon clicking the popover trigger, I see: ```Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.```

One way to clear the error (for debugging purposes) is to not attach the popoverRef by commenting out [line 60](https://codesandbox.io/s/nervous-murdock-fi2x4?file=/src/components/Popover.tsx:1464-1482). The popover will then correctly render, though I'm assuming we also lose a lot of React-ARIA functionality so it isn't a real fix.

| Software         | Version(s) |
| ---------------- | ---------- |
|react-aria   | ^3.10.0
| Browser          | Chrome 94.0.4606.71
| Operating System | Big Sur

**Comment by snowystinger** [comment, marked as answer]
I think if you are seeing it fixed when you do not attach a popoverRef, then you are causing something where the actual node that popoverRef is attached to keeps changing. It's a bit hard to follow all of the render props around, but you can see that you're generating a new node constantly if you put this line of code into your PopoverWithTrigger just after the `useOverlayTrigger`

```
  let prevTrigger = useRef();
  if (prevTrigger.current && triggerRef.current) {
    console.log('trigger', triggerRef.current.isSameNode(prevTrigger.current))
  }
  useEffect(() => {
    if (triggerRef.current) {
      // @ts-ignore
      prevTrigger.current = triggerRef.current;
    }
  }, [triggerRef.current]);
```

I did chase it down to this effect https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/overlays/src/useOverlayPosition.ts#L126, which has a dep on the triggerRef.current

**Reply by bstro** [reply]
Thanks for your assistance @snowystinger! I added that code per your suggestion. [Here's a fork](https://codesandbox.io/s/fragrant-worker-9iqjp?file=/src/components/PopoverWithTrigger.tsx:2174-2254) of the original sandbox with your code snippet in place. Notice the console output, I'm seeing `true` as the value returned from `isSameNode`—doesn't this indicate that I'm not generating a new node?

If it were a different node, `isSameNode` should return `false`. Right?

**Reply by snowystinger** [reply]
Welp, that's confusing, I was seeing false running it locally... But I've pulled this one locally as well and see it's true, I have no difference between the two files. :-/
I'll keep looking, sorry for a red herring

**Reply by snowystinger** [reply]
ah, I tried to fix my own code and stopped it from working instead. useEffect seems to be too delayed, originally i'd written it without that and then thought to myself that looks wrong, but in this case it's actually what we want I'm pretty sure

```
  let prevTrigger = useRef();
  if (prevTrigger.current && triggerRef.current) {
    console.log('trigger', triggerRef.current.isSameNode(prevTrigger.current))
  }
  if (triggerRef.current) {
    // @ts-ignore
    prevTrigger.current = triggerRef.current;
  }
```

https://codesandbox.io/s/great-glitter-rfekk

**Reply by bstro** [reply]
Thank you for your efforts and the helpful hints. I can confirm that the node seems to be replaced resulting in this infinite rendering loop. I was at a loss for what to try, so I went back and re-implemented the whole thing, step-by-step, to make sure I hadn't made any hard-to-catch errors (thats not to suggest that I have any certainty that I have not made any errors), and I'm running into the same error.

https://codesandbox.io/s/reattempt-w-useobjectref-2zwrc?file=/src/App.tsx

I also replaced my ref checks with the upcoming [useObjectRef](https://github.com/adobe/react-spectrum/pull/2293) and that actually changed the behavior of the error a *tiny* bit: instead of receiving `Maximum update depth exceeded` on the initial click, you can now see that the popover displays. However, if you close the popover and reopen, the error occurs.

It is also (maybe) interesting that the positioning doesn't seem to work.

I made [a simpler version](https://codesandbox.io/s/simplified-3zj5u?file=/src/App.tsx) that has fewer render props and it works, though oddly the positioning doesn't work on the first toggle, but corrects itself in subsequent toggles. I'd still really like to get this to work with the original render prop solution. Can you think of any reason why React-ARIA patterns might not play nicely with render props?

I think my next step is to dive into React-ARIA source and throw some debuggers here and there. If you (or anyone) have any ideas, I'd love to hear!

**Reply by snowystinger** [reply]
So good news, I think you've identified a bug in useObjectRef, I'll try to figure out some more information on that to relay to the author.

I think I've boiled it down to the representation of the problem as well as how to solve it, just substitute the commented code out
https://codesandbox.io/s/exciting-lichterman-0wh8b?file=/src/App.js
I haven't tried it on your example yet, but I think it should apply. It's always updating the current ref, because it's generating a new function each time, and the React loses the previous element in the JSX tree because each new function is a new branch where it doesn't know what will be rendered.
I think you'll still have more renders than you would like because of the way that render props work.

**Reply by bstro** [reply]
Thank you immensely, @snowystinger — that was very educational for me. I'm out until Monday, but I'll report in once I give this a shot.

**Reply by bstro** [reply]
Feels like I'm very close.

I can toggle the popover successfully. Only problem is, once I add the 'positionProps' or 'overlayProps' to [the dependency array](https://codesandbox.io/s/somewhat-working-2v1oi?file=/src/PopoverWithTrigger.tsx:2429-2450) inside `PopoverWithTrigger`, the infinite loop returns. I feel like that is a clue, but I don't have enough experience to see the full picture yet. Will continue to try to debug this.

edit: same error also occurs if I try to add `triggerProps` to the [dependency array of Trigger](https://codesandbox.io/s/somewhat-working-2v1oi?file=/src/PopoverWithTrigger.tsx:1955-1981).

**Reply by snowystinger** [reply]
My guess, those hooks return a new object each time. Unfortunately the dependency array is an array, not an object. Which means order matters. And those objects may not always have the same keys, so just sorting the object and spitting it out into an array probably won't work.
That they are a new object each time hasn't mattered because hooks are typically meant to be spread onto a JSX element, in which case React will diff for prop changes. However, we've taken that into our own hands with the useCallback.
You might try something like this (psuedo code, untested)
```
  // Get popover positioning props relative to the trigger
  let { overlayProps: positionProps } = useOverlayPosition({
    targetRef: triggerRef,
    overlayRef,
    placement: "top",
    offset: 5,
    isOpen: state.isOpen
  });
  let oldPositionProps = useRef({});
  let stablePositionProps = useMemo(() => {
    if ( /* use React's algorithm to compare instead of this equals, oldPositionProps.current === positionProps */) {
      return oldPositionProps.current;
    }
    return positionProps;
  }, [positionProps]);
  oldPositionProps.current = stablePositionProps;
  // now use stable instead in your useCallback dep array and for what you spread on the component
```

You can check out React's comparison algorithm https://github.com/facebook/react/blob/996da67b2991f480b6219471e38aeaa7280bd00c/packages/shared/shallowEqual.js#L18

You might be able to reduce some of the above code using something like https://reactjs.org/docs/react-api.html#reactmemo as well

**Reply by snowystinger** [reply]
Just checking in, were you able to get things working?

**Reply by bstro** [reply]
Hey, thanks @snowystinger, I appreciate the check-in! I decided that attempting to use the render-props pattern in the particular way I had been felt like it went too far against the grain of react-aria. Admittedly, I think it also made the code harder to understand overall. Here's a demo of the basic API we ended up implementing: https://codesandbox.io/s/safer-ui-popover-draft-forked-2em8c

The basic strategy was to try and hide as much of the work of wiring all the popover hooks together from consumers of the component. As a result, I ended up with a hook called `usePopoverWithTrigger` that returns the props & refs to pass into both the trigger and popover element. 

Popover's `children` is a render prop that receives `titleProps`. This feels a little leaky, in the sense that it feels like an implementation detail that we couldn't wire up for the consumer, but its not too bad imo. 

If you have any thoughts/suggestions, I'd love to hear—regardless, I'm super grateful for all your help.

**Reply by snowystinger** [reply]
Creating hooks that combine our stately and aria hooks is actually a goal we eventually have in order to minimize the amount of wiring up you need to do. Glad you were able to reach something you are happy with :)
We solved for the title issue with something we refer to as Slots, you can find some info about them (https://react-spectrum.adobe.com/react-spectrum/layout.html#slots), as well as their implementation, which is searchable in the code base. You could do something similar. Essentially it's just using context to pass the props down to the right element(s).
I'm happy to talk pros and cons of it if you go that way and want to know what to look out for.
Here's our RFC that started it as well https://github.com/adobe/react-spectrum/blob/main/rfcs/2019-v3-slots.md

**Reply by bstro** [reply]
I would love to hear about pros and cons to the slots approach, if you're willing to share. I'm going to dive into that slots RFC tomorrow, it sounds promising and we're pretty early on in our component library discussions, so I'm eager to learn as much as I can about different approaches before we start in earnest.

**Reply by snowystinger** [reply]
Of course! I had started working on a blog post for it, but haven't had the time to finish it up yet, you can find its (very) rough draft here https://github.com/adobe/react-spectrum/compare/slots-blog?expand=1
I think at a high level, its pros and cons are as follows
Pros:
- easily send information down into children regardless of nesting and intermediary components
- we treat the slots as just more props, but the parent supplying the slots knows best, so prevents some level of misuse
- some aspects are easy to document, such as just use "these" components
- cuts down on specialty component ie: `Dialog.Content` and `ListItem.Content` can both just be `Content`
- it's not part of anything but the rendering level, so it's not in any of our aria-hooks or stately-hooks
- pairs really nicely with CSS Grid named areas https://css-tricks.com/snippets/css/complete-guide-grid/#grid-template-areas, but works anywhere really

Cons:
- it's context, it'll spread through the entire subtree, so you have to be careful to clear out that context and use it intentionally
- sometimes a bit hard to document as the names for the slots aren't really surfaced, somewhat intentional to prevent context collisions for people using our components, but annoying if you need to supply your own slot name
- rules around it are more like guidelines, can't easily make lint check you
- there's a few places still evolving, such as does the parent always know best? we know with id this isn't the case, which is why this line is the way it is https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/utils/src/Slots.tsx#L26 but are there others where we should always use the child's props instead and use the slots as a default, we haven't found any yet

**Reply by snowystinger** [reply]
@bstro Just wanted to check in and see where you had landed on this topic. Would love to hear more about what you've tried as well.

**Reply by bstro** [reply]
Thanks for checking in, @snowystinger! Other tasks took priority since the last we've spoken on this, and I am just now hopping back on. I read your article on slots, and the approach makes sense to me. I've felt the pain-points of  each approach that you've described in that article. Do you happen to have a simple example of how you implemented the slots architecture? Conceptually it makes sense, but I'd love to see a concrete implementation if you have something nearby/on-hand. Thank you!

**Reply by snowystinger** [reply]
For sure, we have it in React Spectrum :) so we've been developing with it for a while now.

Here's slots themselves, the context
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/utils/src/Slots.tsx

Here's a usage example setting it up
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/dialog/src/Dialog.tsx#L63
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/dialog/src/Dialog.tsx#L97

Here's a component that makes use of slots
https://github.com/adobe/react-spectrum/blob/1e2b7f280a04ead83e21b27fa580eb189e5f4186/packages/%40react-spectrum/view/src/Content.tsx#L21

and here's an example where we expect a lot of content provided by a user so we don't want our slots to continue to propagate
https://github.com/adobe/react-spectrum/blob/1e2b7f280a04ead83e21b27fa580eb189e5f4186/packages/%40react-spectrum/view/src/Content.tsx#L31

**Reply by bstro** [reply]
I've been exploring a bunch of different ideas and have landed on an approach that I'm feeling good about so far. Its basically a combination of tailwindcss+react aria+your slot prop approach (albeit heavily pared-down at the moment).

Here's a draft gist, forking some code from react-spectrum's Slots code. https://gist.github.com/bstro/65e7f5b3fb7743aa1fafe0d84657fdba

Here's a heavily abridged version of a hook that uses the above slot props implementation: https://gist.github.com/bstro/a3c4f0771635625b24322cf1e635b8ec

Here's a `ButtonGroup` component where the slot context is being prepared (see example in the followup comment below this): https://gist.github.com/bstro/90e4283afb7a8d11f084980d905a0540

This approach leverages a library called [`tailwindcss-classnames`](https://github.com/muhammadsammy/tailwindcss-classnames) which provides type safety while using tailwindcss classes via the popular [`classnames`](https://github.com/JedWatson/classnames) api. Also, currently using some methods from Spectrum's implementation.

One drawback of my current implementation is that the overrides cannot apply to one variation or another (for instance, what if I want a different style override for the `primary` Button but not the `secondary` Button?) If I noodle on this more, I'm sure I can smooth and improve that aspect of the API.

I'd love to hear your thoughts if/whenever you have a moment :)

**Reply by bstro** [reply]
The result, btw, of the above code examples, is that given the following button with rounded corners on all sides:
<img width="138" alt="image" src="https://user-images.githubusercontent.com/627290/156466802-53d0e67c-7315-4dae-b523-a148d8f0bb6b.png">

I am able to write the following JSX:
```
<ButtonGroup>
  <Button>Foo</Button>
  <Button>Foo</Button>
  <Button>Foo</Button>
</ButtonGroup>
```

which renders as this:
<img width="172" alt="image" src="https://user-images.githubusercontent.com/627290/156466860-8206d3c3-a25e-4d52-8cac-f32c1b499184.png">

As you can see, the `Button`s inside a `ButtonGroup` have the correct borderRadius applied without the caller having to specify anything at all, and no need for any kind of duplication like `<ButtonGroup.Button>` might entail.
